### PR TITLE
changed init and apply functions to return regular dicts instead of FrozenDicts

### LINKED ISSUE #1223: Should Flax return FrozenDicts or regular dicts?
This topic is discussed regularly internally, and I feel we haven't reached a consensus here. Below are some arguments collected from users for both positions, feel free to add.

**Arguments in favor of FrozenDict**

- @avital: If you use normal dicts, it is easy to mutate them, which means the behavior may differ depending on whether the function in which the modification is made is jitted or not. Example:

```python
def f(params):
  params['conv1']['weight'] = ...
  return ...some computation over params

params = load_from_checkpoint()
print(f(params))
# now what is the value of params['conv1']['weight']?
# depending on whether f is jitted or not, you'd get different results
```

**Arguments in favor of regular dicts**

- @lucasb-eyer: Flax tells me "here's these precious weights, please hold them for me and give them back to me later on, but DONT TOUCH" it begs the question: why give them to me in the first place, if I'm not supposed to do anything with it?

- @avital:  I also think it'd be better for Flax to return normal Python dicts, but still use FrozenDict within modules (via the `mutable` argument to `apply`).

### PR DESCRIPTION
**Pull request by chiamp**
Resolves #1223.

TODO: update documentation to reflect `FrozenDict` to regular `dict` change

### CODE REVIEW

**Comment by marcvanzee** [comment]
@chiamp it might be easier to do this change internally since I expect you will have a lot of internal test failures and might need to change a lot of code for client codebases.

File: docs/guides/transfer_learning.md [hunk 1]
```diff
@@ -118,9 +119,9 @@ params = variables['params']
 Since `params` are currently random, the pretrained parameters from `vision_model_vars` have to be transfered to the `params` structure at the appropriate location. This can be done by unfreezing `params`, updating the `backbone` parameters, and freezing the `params` again:
 
 ```{code-cell} ipython3
-from flax.core.frozen_dict import freeze
+from flax.core.frozen_dict import unfreeze, freeze
 
-params = params.unfreeze()
+params = unfreeze(params)
 params['backbone'] = vision_model_vars['params']
 params = freeze(params)
```
Comment (cgarciae) [inline docs/guides/transfer_learning.md:126]: If this change lands I think we can just remove `freeze` and `unfreeze` altogether here.

File: flax/linen/module.py [hunk 2]
```diff
@@ -2085,7 +2085,7 @@ def scope_fn(scope, *args, **kwargs):
 def init(fn: Callable[..., Any], module: Module,
          mutable: CollectionFilter = DenyList('intermediates'),
          capture_intermediates: Union[bool, Callable[[Module, str], bool]] = False,
-         ) -> Callable[..., FrozenVariableDict]:
+         ) -> Callable[..., dict]:
```
Comment (marcvanzee) [inline flax/linen/module.py:2088]:
```suggestion
         ) -> Callable[..., Union[FrozenVariableDict, dict]]:
```

**Comment by chiamp** [comment]
Closing this as we decided to use a feature flag for dict migration (#2896 and #2898)

**Pull request closed without merging**